tools/Debugger/NetworkComparator.cpp

```diff
@@ -0,0 +1,362 @@
+}
+
+NetworkComparatorBase::NetworkComparatorBase(
+    Module &mod, const std::string &referenceBackend,
+    const std::string &testBackend, float numericCmpThreshold,
+    bool dumpTensorsForBadLayer)
+    : numericCmpThreshold_(numericCmpThreshold), inputModule_(&mod),
+      dumpTensorsForBadLayer_(dumpTensorsForBadLayer) {
+  DCHECK(mod.getFunctions().size() == 1)
+      << "Module must have exactly one functions";
+  EERefNet_.setBackendName(referenceBackend);
+  mod.clone(&EERefNet_.getModule());
+  EETestNet_.setBackendName(testBackend);
+  mod.clone(&EETestNet_.getModule());
+}
+
+NetworkComparatorBase::InOutTensors RecursiveLayerComparator::hookAndRun(
+    llvm::StringRef layerName, PlaceholderBindings *bindings, bool isRef) {
+  ExecutionEngine execEngine;
+  execEngine.setBackendName(isRef ? EERefNet_.getBackendName()
+                                  : EETestNet_.getBackendName());
+  inputModule_->clone(&execEngine.getModule());
+  HookedFunction hook =
+      hookNode(execEngine.getSingleFunctionFromModule(), layerName, true);
+  PlaceholderBindings &inferBindings =
+      isRef ? inferBindingsRef_ : inferBindingsCheck_;
+  inferBindings.allocate(execEngine.getModule().getPlaceholders());
+  // Copy the tensors from the bindings to the inference bindings
+  // we feed to the Temp network.
+  for (auto &PH : bindings->pairs()) {
+    auto iPH = inferBindings.getPlaceholderByName(PH.first->getName());
+    inferBindings.get(iPH)->assign(PH.second);
+  }
+  InOutTensors inOutTensors;
+  for (const auto &P : hook.outputs) {
+    Tensor *t = inferBindings.get(P);
+    std::string str = P->getName();
+    DCHECK(t) << "Placeholder not " << str << " found in the bindings\n";
+    inOutTensors.outputs[str] = t;
+  }
+  for (const auto &P : hook.inputs) {
+    Tensor *t = inferBindings.get(P);
+    std::string str = P->getName();
+    DCHECK(t) << "Placeholder " << str << " not found in the bindings\n";
+    inOutTensors.inputs[str] = t;
+  }
+  auto fName = hook.function->getName();
+  execEngine.compile(CompilationMode::Infer);
+  execEngine.run(inferBindings, fName);
+  return inOutTensors;
+}
+
+bool RecursiveLayerComparator::verify(PlaceholderBindings *bindings) {
+  bool allPassed = true;
+  // Sort the nodes in topological order to test nodes in that order.
+  // Tensors flow through the network in topological order, testing the layers
+  // in that order will allow us to see how errors propagate.
+  GraphPostOrderVisitor visitor(*EERefNet_.getSingleFunctionFromModule());
+  llvm::ArrayRef<Node *> order = visitor.getPostOrder();
+  for (auto const *nodePtr : order) {
+    auto &node = *nodePtr;
+    DCHECK(nodePtr) << "Node is empty!";
+    if (llvm::isa<SaveNode>(&node) || llvm::isa<Constant>(&node) ||
+        llvm::isa<Placeholder>(&node)) {
+      continue;
+    }
+    std::string layerName = node.getName();
+    std::string kName = node.getKindName();
+    LOG(INFO) << "Verifying layer: " << layerName << "\tType: " << kName
```

Comment:
nit: Not a huge deal, but if you use `llvm::StringRef` above you can use `layerName.data()` here, and we'd never need to copy the string.